### PR TITLE
IGNITE-17056 Design for rebalance cancellation

### PR DESCRIPTION
**Pull request by kgusakov**
https://issues.apache.org/jira/browse/IGNITE-17056

### CODE REVIEW

File: modules/distribution-zones/tech-notes/rebalance.md [hunk 1]
```diff
@@ -55,18 +55,55 @@ But for the new one we have an idea, which doesn't need the metastore at all:
...
+        zoneId.assignment.cancel = cancelValue
+        return true
+    else:
+        return false
+```
+It's needed to prevent the race, between the rebalance done and cancel persisting, otherwise we can try to cancel the wrong rebalance process.
 
+### 2. PrimaryReplica->ReplicationGroup cancel protocol
+When PrimaryReplica send `CancelRebalanceRequest(oldTopology, newTopology)` to the ReplicationGroup following cases available:
+- Replication group has ongoing rebalance oldToplogy->newTopology. So, it must be cancelled and cleanup for the configuration state of replication group to  oldTopology must be executed.
```
Comment (denis-chudov) [inline modules/distribution-zones/tech-notes/rebalance.md:107]:
```suggestion
- Replication group has ongoing rebalance oldTopology->newTopology. So, it must be cancelled and cleanup for the configuration state of replication group to oldTopology must be executed.
```

File: modules/distribution-zones/tech-notes/rebalance.md [hunk 2]
```diff
@@ -55,18 +55,55 @@ But for the new one we have an idea, which doesn't need the metastore at all:
...
+- Replication group has no ongoing rebalance and currentTopology==oldTopology. So, nothing to cancel, return success response.
+- Replication group has no ongoing rebalance and currentTopology==newTopology. So, cancel request can't be executed, return the response about it. Result recipient of this response (placement driver) must log this fact and do the same routine for usual rebalanceDone.
```
Comment (denis-chudov) [inline modules/distribution-zones/tech-notes/rebalance.md:109]: What if after sending `CancelRebalanceRequest` the placement driver finds out that some of replication groups have finished rebalance (currentTopology==newTopology) and some have not?

File: modules/distribution-zones/tech-notes/src/cancelRebalance.puml [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+ReplicationGroup -> PrimaryReplica : Send rebalanceError message
+PrimaryReplica -> PlacementDriver : Send rebalanceError message
+PlacementDriver -> PlacementDriver: Decides to give up with this rebalance at all
+PlacementDriver -> Metastore : Put (oldTopology, newTopology) to **zonedId.assignment.cancel** key [see 1]
+PlacementDriver <- Metastore : Receives the notification about ***.cancel** update
+PlacementDriver -> PrimaryReplica : Send CancelRebalanceRequest(oldTopology, newTopology)
+PrimaryReplica -> ReplicationGroup : Send CancelRebalanceRequest(oldTopology, newTopology) [see 2]
+ReplicationGroup -> ReplicationGroup : Process cancellation request and invoke onCancelDone listener
+ReplicationGroup -> PrimaryReplica : Send cancelnRebalanceDone message
+PrimaryReplica -> PlacementDriver : Send cancelnRebalanceDone message
```
Comment (denis-chudov) [inline modules/distribution-zones/tech-notes/src/cancelRebalance.puml:15]: `cancelnRebalanceDone` is the name correct? You mean `CancelRebalanceDone`?

**Review by denis-chudov — CHANGES_REQUESTED** [review]

File: modules/distribution-zones/tech-notes/rebalance.md (hunk 2 above)
Comment (kgusakov) [reply modules/distribution-zones/tech-notes/rebalance.md:109]: How does it possible, if any rebalance touches only one distribution zone, so, only one replication group?

Comment (denis-chudov) [reply modules/distribution-zones/tech-notes/rebalance.md:109]: Seems I missed this moment, my fault. Question withdrawn.

File: modules/distribution-zones/tech-notes/rebalance.md [hunk 4]
```diff
@@ -55,18 +55,62 @@ But for the new one we have an idea, which doesn't need the metastore at all:
...
+- On every new PD leader elected - it must check the direct value (not the locally cached one) of `zoneId.assignment.pending`/`zondeId.assignment.cancel` (the last one always wins, if exists) keys and send `RebalanceRequest`/`CancelRebalanceRequest` to needed PrimaryReplicas and then listen updates from the last revision of this key.
+- On every PrimaryReplica reelection by PD it must send the new `RebalanceRequest`/`CancelRebalanceRequest` to PrimaryReplica, if pending/cancel (cancel always wins, if filled) key is not empty. 
+- On every leader reelection (for the leader oriented protocols) inside the replication group - leader sends leaderElected event to PrimaryReplica, which forces PrimaryReplica to send RebalanceRequest/CancelRebalanceRequest to the replication group leader again.
+
+More over: 
+- `RebalanceRequest`/`CancelRebalanceRequest` must include the revision of its' trigger. 
+- PrimaryReplica must persist the last seen revision locally.
+- When new PrimaryReplica elected, PlacementDriver must initialize the last seen revision of PrimaryReplica to the current revision-1. So, after that PlacementDriver must send the *Request with current actual revision.
```
Comment (denis-chudov) [inline modules/distribution-zones/tech-notes/rebalance.md:89]:
```suggestion
- When new PrimaryReplica elected, PlacementDriver must initialize the last seen revision of PrimaryReplica to the `currentRevision-1`. So, after that PlacementDriver must send the *Request with current actual revision.
```

File: modules/distribution-zones/tech-notes/rebalance.md [hunk 5]
```diff
@@ -55,18 +55,62 @@ But for the new one we have an idea, which doesn't need the metastore at all:
...
+
+### 5. Stop the redundant replicas and update replicas clients
 Here we need to:
 - Stop the redundant replicas, which is not in the current stable assignments
   - We can accidentally stop PrimaryReplica, so we need to use the algorithm of a graceful PrimaryReplica transfer, if needed.
 - Update the replication protocol clients (RaftGroupService, for example) on each Replica.
 
-### Failover logic
+## Failover logic
 The main idea of failover process: every rebalance request PlacementDriver->PrimaryReplica or PrimaryReplica->ReplicationGroup must be idempotent. So, redundant request in the worst case should be just answered by positive answer (just like rebalance is already done).
```
Comment (denis-chudov) [inline modules/distribution-zones/tech-notes/rebalance.md:79]:
```suggestion
The main idea of failover process: every rebalance request and cancel rebalance request PlacementDriver->PrimaryReplica or PrimaryReplica->ReplicationGroup must be idempotent. So, redundant request in the worst case should be just answered by positive answer (just like rebalance is already done).
```

**Review by denis-chudov — CHANGES_REQUESTED** [review]

**Review by denis-chudov — APPROVED** [review]

**Pull request closed without merging**